internal/luks2/activate.go

```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package luks2
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+
+	"golang.org/x/xerrors"
+)
+
+var (
+	systemdCryptsetupPath = "/lib/systemd/systemd-cryptsetup"
+)
+
+// Activate unlocks the LUKS device at sourceDevicePath using systemd-cryptsetup and creates a device
+// mapping with the supplied volumeName. The device is unlocked using the supplied key.
+//
+// This will return a *os.ExitError error in the event that systemd-cryptsetup fails.
+func Activate(volumeName, sourceDevicePath string, key []byte) error {
+	fifoPath, cleanupFifo, err := mkFifo()
+	if err != nil {
+		return xerrors.Errorf("cannot create FIFO for passing key to systemd-cryptsetup: %w", err)
+	}
+	defer cleanupFifo()
+
+	cmd := exec.Command(systemdCryptsetupPath, "attach", volumeName, sourceDevicePath, fifoPath, "luks,tries=1")
```

Comment:
Does `-` as input work with systemd-cryptsetup? In which case you could probably avoid using fifo and hook up a bytes.Buffer with the key to cmd.Stdin.

If `-` isn't supported, what may be a bit hackish, I think one could specify /dev/stdin as the input 'file'.

Reply:
Ah, passing /dev/stdin as the path seems to work - I verified this manually. I've updated the code to do this instead.

Note that systemd-cryptsetup is mocked in the tests, so I still need to verify that the updated implementation works correctly with the real systemd-cryptsetup.